src/v2/cookbook/unit-testing-vue-components.md

```diff
@@ -0,0 +1,234 @@
+  </div>
+</template>
+
+<script>
+export default {
+  name: 'Foo',
+
+  data () {
+    return {
+      message: 'Hello World',
+      username: ''
+    }
+  },
+
+  computed: {
+    error () {
+      return this.username.length < 7
+    }
+  }
+}
+</script>
+```
+
+The things that we should test are:
+- is the `message` rendered?
+- if `error` is `true`, `<div class="error"`> should be present
+- if `error` is `false`, `<div class="error"`> should not be present
+
+And our first attempt at at test:
+
+```js
+import { shallow } from 'vue-test-utils'
+
+describe('Foo', () => {
+  it('renders a message and responds correctly to user input', () => {
+      const wrapper = shallow(Foo, {
+    data: {
+      message: 'Hello World',
+      username: ''
+    }
+  })
+
+  // see if the message renders
+  expect(wrapper.find('.message').text()).toEqual('Hello World')
+
+  // assert the error is rendered
+  expect(wrapper.find('.error').exists()).toBeTruthy()
+
+  // update the username and assert error is longer rendered
+  wrapper.setData({ username: 'Lachlan'  })
+  expect(wrapper.find('.error').exists()).toBeFalsey()
+  })
+})
+```
+
+There are some problems with the above:
+- a single test is making assertions about different things
+- difficult to tell the different states the component can be in, and what should be rendered
+
+The below example improves the test by:
+- only making one assertion per `it` block
+- having short, clear test descriptions
+- providing only the minimum data requires for the test
+
+*Updated test*:
+```js
+import { shallow } from 'vue-test-utils'
+
+describe('Foo', () => {
+  it('renders a message', () => {
+    const wrapper = shallow(Foo, {
+      data: {
+        message: 'Hello World'
+      }
+    })
+
+    expect(wrapper.find('.message').text()).toEqual('Hello World')
+  })
+
+  it('renders error when username is less than 7 characters', () => {
+    const wrapper = shallow(Foo, {
+      data: {
+        username: ''
+      }
+    })
+
+    expect(wrapper.find('.error').exists()).toBeTruthy()
+  })
+
+  it('does not render error when username is 7 characters or more', () => {
+    const wrapper = shallow(Foo, {
+      data: {
+        username: 'Lachlan'
+      }
+    })
+
+    expect(wrapper.find('.error').exists()).toBeFalsey()
+  })
+})
+```
+
+## Additional Context
+
+Thee above test is fairly simple, but in practise Vue components often have other behaviors you want to test, such as:
+
+- making API calls
+- committing or dispatching mutations or actions with a `Vuex` store
+- testing interaction
+
+`vue-test-utils` and the enormous JavaScript ecosystem provides plenty of tooling to facilitate almost 100% test coverage. Unit tests are only one part of the testing pyramid, though. Some other types of tests include e2e (end to end) tests, and snapshot tests. Unit tests are the smallest and most simple of tests - they make assertions on the smallest units of work, isolating each part of a single component. 
```

Comment:
It might be worth mentioned what e2e and snapshot tests are in contrast to unit tests (only a very short description is needed)